lib/statsd.js

```diff
@@ -526,3 +502,59 @@ Client.prototype.childClient = function (options) {
 
 exports = module.exports = Client;
 exports.StatsD = Client;
+
+function udsErrorHandler(client, lastSocketCreateTime) {
+  // recreate the socket, but only once within `udsGracefulRestartRateLimit`.
+  if (Date.now() - lastSocketCreateTime < client.udsGracefulRestartRateLimit) {
+    return;
+  }
+
+  if (client.errorHandler) {
+    client.socket.removeListener("error", client.errorHandler);
+  }
+
+  const newSocket = createTransport(client, {
+    host: client.host,
+    path: client.path,
+    port: client.port,
+    protocol: client.protocol,
+  });
+  if (newSocket) {
+    client.socket.close();
+    client.socket = newSocket;
+    maybeAddUDSErrorHandler(this, Date.now());
+  } else {
+    console.error('Could not replace UDS connection with new socket');
+    return;
+  }
+
+  if (client.errorHandler) {
+    client.socket.on("error", client.errorHandler);
+  } else {
+    client.socket.on("error", (error) => console.error(`hot-shots UDS error: ${error}`));
+  }
+}
+
+function maybeAddUDSErrorHandler(client, lastSocketCreateTime) {
```

Comment:
I'm very open to a naming convention other than `maybe` in `maybeAddUDSErrorHandler()`

Reply:
I don't get too hung up on names, unless I think it doesn't fit the style.  Not too much of a naming style in here so it's fine.  :)  Spent a bunch of time making sure I understood the new paths in here, for this function and others.  I didn't see any issues, looks good to me.